demo/c/rhino_demo_mic.c

```diff
@@ -286,140 +191,217 @@ int main(int argc, char *argv[]) {
-    device_config.capture.pDeviceID = &capture_info[device_index].id;
-    device_config.capture.channels = 1;
-    device_config.sampleRate = pv_sample_rate_func();
-    device_config.dataCallback = mic_callback;
-    device_config.pUserData = &pv_rhino_data;
-
-    result = ma_device_init(&context, &device_config, &device);
-    if (result != MA_SUCCESS) {
-        fprintf(stderr, "failed to initialize capture device.\n");
+    const char *context_info = NULL;
+    status = pv_rhino_context_info_func(rhino, &context_info);
+    if (status != PV_STATUS_SUCCESS) {
+        fprintf(stderr, "'pv_rhino_context_info' failed with '%s'\n", pv_status_to_string_func(status));
         exit(1);
     }
+    fprintf(stdout, "%s\n\n", context_info);
 
-    result = ma_device_start(&device);
-    if (result != MA_SUCCESS) {
-        fprintf(stderr, "failed to start device.\n");
+    const char *selected_device = pv_recorder_get_selected_device(recorder);
+    fprintf(stdout, "Selected device: %s.\n", selected_device);
+    fprintf(stdout, "Listening...\n\n");
+
+    recorder_status = pv_recorder_start(recorder);
+    if (recorder_status != PV_RECORDER_STATUS_SUCCESS) {
+        fprintf(stderr, "Failed to start device with %s.\n", pv_recorder_status_to_string(recorder_status));
         exit(1);
     }
 
-    fprintf(stdout, "Using device: %s\n", device.capture.name);
-    fflush(stdout);
-
-    const char *context_info = NULL;
-    status = pv_rhino_context_info_func(rhino, &context_info);
-    if (status != PV_STATUS_SUCCESS) {
-        fprintf(stderr, "'pv_rhino_context_info' failed with '%s'\n", pv_status_to_string_func(status));
+    int16_t *pcm = malloc(frame_length * sizeof(int16_t));
+    if (!pcm) {
+        fprintf(stderr, "Failed to allocate pcm memory.\n");
         exit(1);
     }
-    fprintf(stdout, "%s\n\n", context_info);
-    fflush(stdout);
 
-    while (!is_done);
+    while (!is_interrupted) {
 
-    ma_device_uninit(&device);
-    ma_context_uninit(&context);
+        recorder_status = pv_recorder_read(recorder, pcm);
+        if (recorder_status != PV_RECORDER_STATUS_SUCCESS) {
+            fprintf(stderr, "Failed to read with %s.\n", pv_recorder_status_to_string(recorder_status));
+            exit(1);
+        }
 
-    free(pv_rhino_data.frame_buffer.buffer);
-    pv_rhino_delete_func(rhino);
+        bool is_finalized = false;
+        status = pv_rhino_process_func(
+                rhino,
+                pcm,
+                &is_finalized);
+
+        if (status != PV_STATUS_SUCCESS) {
+            fprintf(stderr, "'pv_rhino_process' failed with '%s'\n", pv_status_to_string_func(status));
+            exit(1);
+        }
+
+        if (is_finalized) {
+            bool is_understood = false;
+            status = pv_rhino_is_understood_func(rhino, &is_understood);
+            if (status != PV_STATUS_SUCCESS) {
+                fprintf(stderr, "'pv_rhino_is_understood' failed with '%s'\n",
```

Comment:
we do some clean up at the end of this program, but what if something goes wrong? the program just exits with an error code; i am not sure what's the best thing to do here to be honest since Linux (and probably mac) is doing a good job cleaning everything after an application exits.